ERCS/erc-7820.md

```diff
@@ -0,0 +1,411 @@
+    event ContractUnregistered(address indexed _contract, address indexed _admin);
+
+    // Emitted when a role is granted to an account for a contract.
+    // @param targetContract The address of the contract.
+    // @param role The role being granted.
+    // @param account The address of the account.
+    event RoleGranted(
+        address indexed targetContract,
+        bytes32 indexed role,
+        address indexed account
+    );
+
+    // Emitted when a role is revoked from an account for a contract.
+    // @param targetContract The address of the contract.
+    // @param role The role being revoked.
+    // @param account The address of the account.
+    event RoleRevoked(
+        address indexed targetContract,
+        bytes32 indexed role,
+        address indexed account
+    );
+
+    // Registers a contract with the given admin.
+    // @param _admin The address of the admin for the registered contract.
+    function registerContract(address _admin) external;
+
+    // Unregisters a contract.
+    // @param _contract The address of the contract to unregister.
+    function unRegisterContract(address _contract) external;
+
+    // Grants roles to multiple accounts for multiple contracts.
+    // @param targetContracts An array of contract addresses to which roles will be granted.
+    // @param roles An array of roles to be granted.
+    // @param accounts An array of accounts to be granted the roles.
+    function grantRole(
+        address[] memory targetContracts,
+        bytes32[] memory roles,
+        address[] memory accounts
+    ) external;
+
+    // Revokes roles from multiple accounts for multiple contracts.
+    // @param targetContracts An array of contract addresses from which roles will be revoked.
+    // @param roles An array of roles to be revoked.
+    // @param accounts An array of accounts from which the roles will be revoked.
+    function revokeRole(
+        address[] memory targetContracts,
+        bytes32[] memory roles,
+        address[] memory accounts
+    ) external;
+
+    
+    //Gets the information of a registered contract.
+    //@param _contract The address of the contract to get the information.
+    //@return isActive Whether the contract is active.
+    //@return admin The address of the admin for the contract.
+    //MUST revert if the registered contract doesn't exist
+    function getContractInfo(
+        address _contract
+    ) external view returns (bool isActive, address admin);
+
+    // Gets the information of a registered contract.
+    // @param _contract The address of the contract to get the information.
+    // @return isActive Whether the contract is active.
+    // @return admin The address of the admin for the contract.
+    // MUST revert if the registered contract doesn't exist`
+    function getRoleInfo(
+        address _contract
+    ) external view returns (bool isActive, address admin);
+}
+
+```
+
```

Comment:
I think this could benefit from a bit of "usage" text. Like who is allowed to grant/revoke roles, and who registers the contract? I assume the contract registers itself, and the admin address is allowed to grant roles, but you should make that explicit in this section.

Your reference implementation is meant to supplement the specification section. Implementers shouldn't _have_ to dig into it to figure out what the standard is trying to say. The Specification section should have all the information on its own.